protocols/imap/src/main/java/org/apache/james/imap/main/PathConverter.java

```diff
@@ -21,80 +21,230 @@
-        if (mailboxName.equalsIgnoreCase(MailboxConstants.INBOX)) {
-            return MailboxConstants.INBOX;
+        /**
+         * Joins the elements of a mailboxPath together and returns them as a string
+         */
+        private String joinMailboxPath(MailboxPath mailboxPath, MailboxSession session) {
+            StringBuilder sb = new StringBuilder();
+            if (mailboxPath.getNamespace() != null && !mailboxPath.getNamespace().isEmpty()) {
+                if (mailboxPath.getNamespace().equalsIgnoreCase(MailboxConstants.USER_NAMESPACE)
+                    && !mailboxPath.belongsTo(session)) {
+                    sb.append("#user");
+                } else {
+                    sb.append(mailboxPath.getNamespace());
+                }
+            }
+            if (mailboxPath.getUser() != null) {
+                if (!mailboxPath.belongsTo(session)) {
+                    if (!sb.isEmpty()) {
+                        sb.append(session.getPathDelimiter());
+                    }
+
+                    sb.append(USERNAME_ESCAPER.escape(mailboxPath.getUser().getLocalPart()));
+                }
+            }
+            if (mailboxPath.getName() != null && !mailboxPath.getName().isEmpty()) {
+                if (!sb.isEmpty()) {
+                    sb.append(session.getPathDelimiter());
+                }
+                sb.append(mailboxPath.getName());
+            }
+            return sb.toString();
+        }
+
+        public String mailboxName(boolean relative, MailboxPath path, MailboxSession session) {
+            if (relative && path.belongsTo(session)) {
+                return path.getName();
+            } else {
+                return joinMailboxPath(path, session);
+            }
+        }
+
+        public MailboxQuery mailboxQuery(String finalReferencename, String mailboxName, ImapSession session) {
+            MailboxSession mailboxSession = session.getMailboxSession();
+            String decodedMailboxName = ModifiedUtf7.decodeModifiedUTF7(mailboxName);
+            if (finalReferencename.isEmpty()) {
+                if (mailboxName.equals("*")) {
+                    return MailboxQuery.builder()
+                        .matchesAllMailboxNames()
+                        .build();
+                }
+                return MailboxQuery.builder()
+                    .expression(new PrefixedRegex(
+                        "",
+                        decodedMailboxName,
+                        mailboxSession.getPathDelimiter()))
+                    .build();
+            }
+
+            MailboxPath basePath = computeBasePath(session, finalReferencename);
+            if (basePath.getNamespace().equals(MailboxConstants.USER_NAMESPACE)
+                && basePath.getUser() == null) {
+
+                int separatorPosition = decodedMailboxName.indexOf(mailboxSession.getPathDelimiter());
+                if (separatorPosition >= 0) {
+                    // interpret first part as the user
```

Comment:
If we use `.split()` we won't need that comment.

Reply:
:-1: split underneath uses a regex which obfuscate code, expose to complications like REDOS, and behaves pourly in what is here a critical path.